ts/metrics/metric_cache_yaml_impl.py

```diff
@@ -0,0 +1,198 @@
+            )
+
+    def _parse_metrics_section(self, key="model_metrics") -> dict:
+        """
+        Given a key present in the yaml, returns the corresponding section
+
+        Parameters
+        ----------
+        key: str
+            section of yaml file to be parsed
+        """
+        try:
+            val = self._parsed_file[key]
+        except KeyError as err:
+            raise merrors.MetricsCacheKeyError(
+                f"'{key}' key not found in yaml file: {err}"
+            )
+        logging.debug(f"Successfully parsed {key} section of yaml file")
+        return val
+
+    def initialize_cache(self) -> None:
+        """
+        Create Metric objects based off of the model_metrics data and add to cache
+        """
+        metrics_section = self._parse_metrics_section("model_metrics")
+        if not metrics_section:
+            raise ValueError("Missing `model_metrics` specification")
+        for metric_type, metrics_list in metrics_section.items():
+            try:
+                metric_enum = MetricTypes(metric_type)
+            except Exception as exc:
+                raise merrors.MetricsCacheKeyError(f"Invalid metric type: {exc}")
+
+            for metric in metrics_list:
+                try:
+                    metric_name = metric["name"]
+                    unit = metric["unit"]
+                    dimension_names = metric["dimensions"]
+                    self.add_metric(
+                        metric_name=metric_name,
+                        unit=unit,
+                        dimension_names=dimension_names,
+                        metric_type=metric_enum,
+                    )
+                except KeyError as k_err:
+                    raise merrors.MetricsCacheKeyError(f"Key not found in cache spec: {k_err}")
+
+    def add_metric(
+        self,
+        metric_name: str,
+        unit: str,
+        dimension_names: list = [],
+        metric_type: MetricTypes = MetricTypes.COUNTER,
+    ) -> CachingMetric:
+        """
+        Create a new metric and add into cache
+
+        Parameters
+        ----------
+        metric_name str
+            Name of metric
+        unit str
+            unit can be one of ms, percent, count, MB, GB or a generic string
+        dimension_names list
+            list of dimension keys which should be strings, or the complete log of dimensions
+        metric_type MetricTypes
+            Type of metric Counter, Gauge, Histogram
+        Returns
+        -------
+        newly created Metrics object
+        """
+        if (
+            not isinstance(metric_name, str)
+            or not isinstance(unit, str)
+            or not isinstance(metric_type, MetricTypes)
+        ):
+            raise merrors.MetricsCacheTypeError(
+                f"metric_name must be a str, unit must be a str, "
+                f"dimensions should be a list of Dimension objects/None, "
```

Comment:
Dimension is not checked here. Can we create a method called check_type(variable, expected_type, helper_text) that checks and emits the exception? By unbundling user also know what exactly had the wrong type.

Reply:
Added method `check_type`